src/strategy.ts

```diff
@@ -89,16 +84,22 @@ export abstract class AbstractStrategy extends PassportStrategy {
             }
 
             const RelayState = req.query?.RelayState || req.body?.RelayState;
-            getLogoutResponseUrl(profile, RelayState, options, userMatch, redirectIfSuccess);
-
-            // Log out the current user no matter if we can verify the logged in user === logout requested user
-            req.logout();
-
-            if (!userMatch) {
+            if (this._saml == null) {
               return this.error(
-                new Error("Attempting to log out a user that differs from the current user.")
+                new Error("Can't get logout response URL without a SAML provider defined.")
+              );
+            } else {
+              this._saml.getLogoutResponseUrl(
+                profile,
+                RelayState,
+                options,
+                userMatch,
+                redirectIfSuccess
               );
             }
+
+            // Log out the current user no matter if we can verify the logged in user === logout requested user
+            req.logout();
```

Comment:
I'm not sure about control flow but based on quick code browsing I think that if implementation of `redirectIfSuccess` callback https://github.com/node-saml/passport-saml/blob/591a1d83b8ebfb2c19a2e393b37238974687d090/src/strategy.ts#L135-L143 gets executed by implementation of `getLogoutResponseUrl` this `req.logout()` line is not reached and `passport-saml` would once again return `Success` even if session was not terminated. Although this time reason would be that
1. even though caller was identified - i.e. his/her browser didn't block 3rd party cookie - and
2. this PR's code matched `req.user` with `LogoutRequest`'s information meaning that it was determined that it is a-ok to return `Success` but
3. redirect with `LogoutResponse` with `Success` status was triggered and request processing ended  before reaching this line

Your pull request seems to contain testcases i.e. TDD approach was your goal. I haven't inspected test case changes but based on already closed review findings (issues weren't captured by testcases) it seems that there aren't any tests which would replicate situation from end user's UE point of view / signalling between end user's UE and express.js application at SAML protocol level / HTTP request cookie / actual resource authorization level.

I attached to #419 's description two testcases which try to verify (and at that particular case pin point issues) `passport-saml`'s behaviour at SAML/HTTP protocol level along with enablers for session management tests. 

BTW. given the various possibilities to configure express-session to e.g. update session content after request processing to session store and various session store implementations is there any chance to make a decision - by passport-saml - that content of `session.passport` for given `sessionID` was successfully cleared from session store because:
Assume that
1. developer has configured express session to write session back to session store after each  request
2. application is some SPA application running at end user's browser tab
3. SPA application uses e.g. long polling to get some updates from backend
4. end user triggers SLO from some application B (running at another tab or browser window) which uses same SSO with aforementioned SPA application
5. IdP coordinates `LogoutRequest` to aforementioned SPA application's backend (this coordination happens at tab/browser window used by application B)
6. SPA application had ongoing long polling (which started prior to SLO and ends after SLO has ended)

now...if express session is configured to always update `session` to session store after `req` is processed doesn't that mean that express/nodejs instance which was handling concurrently aforementioned long polling request write content of `session` (along with `session.passport.user`) from `req` which represented long polling request back to session store. I.e. removed session would be reintroduced to session store?

If that concurrent request handling scenario is possible then `passport-saml` would have to have its own book keeping of login sessions (i.e. session would be added to bookkeeping only once during successfull authentication) and removed on normal logout or SLO `LogoutRequest` processing and there would not be any chance to mess things up with uncontrollable/wild 3rd party sw/configurations...and passport-saml's own bookkeeping would need to be checked during each request if request is related to resource which must not be accessible for unauthenticated users.

Sorry if it was hard to interpret this comment (grammar etc.). Due to some reason it was harder than usual to write english.

Reply:
I agree with your scenario, however, the assumption in (1) is really not passport-saml's problem. The developer shouldn't be adding things to a user store just because a request completed.

Since it is out of scope for passport-saml to manage a user store, there has to be some onus on the developer to do this correctly. While there may be opportunities to use passport-saml incorrectly, my goal here is to make sure that it doesn't do that _by default_ as your previous bug reported pointed out.